src/control/ros/Conversions.cpp

```diff
@@ -0,0 +1,438 @@
+  Eigen::VectorXd* _accelerations, bool _accelerationsRequired,
+  std::map<size_t, size_t> indexMap)
+{
+  const auto& waypoint = _trajectory.points[_index];
+
+  try
+  {
+    Eigen::VectorXd trajPos, trajVel, trajAccel;
+    checkVector("positions", waypoint.positions, _numDofs,
+      _positionsRequired, &trajPos);
+    checkVector("velocities", waypoint.velocities, _numDofs,
+      _velocitiesRequired, &trajVel);
+    checkVector("accelerations", waypoint.accelerations, _numDofs,
+      _accelerationsRequired, &trajAccel);
+
+    if ( trajPos.size() > 0 )
+      reorder(indexMap, trajPos, _positions);
+    if ( trajVel.size() > 0 )
+      reorder(indexMap, trajVel, _velocities);
+    if ( trajAccel.size() > 0 )
+      reorder(indexMap, trajAccel, _accelerations);
+  }
+  catch (const std::invalid_argument& e)
+  {
+    std::stringstream message;
+    message << "Waypoint " << _index << " is invalid: " << e.what();
+    throw std::invalid_argument(message.str());
+  }
+}
+
+//=============================================================================
+void extractTrajectoryPoint(
+  const std::shared_ptr<MetaSkeletonStateSpace>& space,
+  const aikido::trajectory::TrajectoryPtr& trajectory,
+  double timeFromStart, const std::map<size_t, size_t>& indexMap,
+  trajectory_msgs::JointTrajectoryPoint& waypoint)
+{
+  const auto numDerivatives = std::min<int>(trajectory->getNumDerivatives(), 1);
+  const auto timeAbsolute = trajectory->getStartTime() + timeFromStart;
+  
+  Eigen::VectorXd tangentVector, jointTangentVector;
+  auto state = space->createState();
+  trajectory->evaluate(timeAbsolute, state);
+  space->logMap(state, tangentVector);
+  assert(tangentVector.size() == numDofs);
+
+  // Reorder tangentVectoraccording to the index mapping
+  reorder(indexMap, tangentVector, &jointTangentVector);
+
+  waypoint.time_from_start = ::ros::Duration(timeFromStart);
+  waypoint.positions.assign(jointTangentVector.data(),
+    jointTangentVector.data() + jointTangentVector.size());
+
+  assert(0 <= numDerivatives && numDerivatives <= 2);
+  const std::array<std::vector<double>*, 2> derivatives{
+    &waypoint.velocities, &waypoint.accelerations};
+  for (int iderivative = 1; iderivative <= numDerivatives; ++iderivative)
+  {
+    trajectory->evaluateDerivative(timeAbsolute, iderivative, tangentVector);
+    assert(tangentVector.size() == numDofs);
+
+    // Reorder according to the index mapping
+    reorder(indexMap, tangentVector, &jointTangentVector);
+    derivatives[iderivative - 1]->assign(jointTangentVector.data(),
+      jointTangentVector.data() + jointTangentVector.size());
+  }
+}
+
+//=============================================================================
+// The rows of inVector is reordered in outVector.
+void reorder(std::map<size_t, size_t> indexMap,
+  const Eigen::VectorXd& inVector, Eigen::VectorXd* outVector)
+{
+  *outVector = Eigen::VectorXd::Zero(inVector.size());
```

Comment:
This is a helper function in an anonymous namespace, so I suggest just adding an `assert`.